src/components/Menu/Menu.tsx

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2023 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { FC, ReactNode, useMemo } from "react";
+import { Root, Trigger, Portal, Content } from "@radix-ui/react-dropdown-menu";
+import { FloatingMenu } from "./FloatingMenu";
+import { Drawer } from "vaul";
+import classnames from "classnames";
+import drawerMenu from "./DrawerMenu.module.css";
+import { MenuContext } from "./MenuContext";
+import { DrawerMenu } from "./DrawerMenu";
+import { useTouchscreen } from "../../utils/useTouchscreen";
+
+interface Props {
+  /**
+   * The menu title.
+   */
+  title: string;
+  /**
+   * Whether the menu is open.
+   */
+  open: boolean;
+  /**
+   * Event handler called when the open state of the menu changes.
+   */
+  onOpenChange: (open: boolean) => void;
+  /**
+   * The button that opens the menu. This must be a component that accepts a ref
+   * and spreads props.
+   * https://www.radix-ui.com/primitives/docs/guides/composition
+   */
+  trigger: ReactNode;
+  /**
+   * The menu contents.
+   */
+  children: ReactNode;
+  /**
+   * The side of the trigger on which to place the menu. Note that the menu may
+   * still end up on a different side than the one you request if there isn't
+   * enough space.
+   * @default bottom
+   */
+  side?: "top" | "right" | "bottom" | "left";
+  /**
+   * The edge along which the menu and trigger will be aligned.
+   * @default center
+   */
+  align?: "start" | "center" | "end";
+}
+
+/**
+ * A menu opened by pressing a button.
+ */
+export const Menu: FC<Props> = ({
+  title,
+  open,
+  onOpenChange,
+  trigger,
+  children: childrenProp,
+  side = "bottom",
+  align = "center",
+}) => {
+  // Normally, the menu takes the form of a floating box. But on mobile (which
+  // we detect by looking for whether the primary input device is a
+  // touchscreen), the menu should morph into a drawer
+  const touchscreen = useTouchscreen();
+  const context = useMemo(
+    () => ({
+      component: touchscreen
+        ? ("Vaul drawer" as const)
+        : ("Radix dropdown menu" as const),
+      onOpenChange,
+    }),
+    [touchscreen, onOpenChange],
+  );
```

Comment:
I'm not sure I understand this piece of code? A device is unlikely to morph from a touch device to a non touch device?

I would also suggest to remove references to `radix` or `vaul` from the context values, as this will not pass the test of time, and will become hard to grok the day we refactor to use another library.